src/openms/include/OpenMS/DATASTRUCTURES/QTCluster.h

```diff
@@ -119,115 +118,97 @@ namespace OpenMS
-     *
-     * For each input run, a multimap which contains pointers to all
-     * neighboring elements and the respective distance.
-     *
-     */
-    NeighborMapMulti* tmp_neighbors_;
-
-    /// Maximum distance of a point that can still belong to the cluster
-    double max_distance_;
-
-    /// Number of input maps
-    Size num_maps_;
-
-    /// Quality of the cluster
-    double quality_;
-
-    /// Has the cluster changed (if yes, quality needs to be recomputed)?
-    bool changed_;
-
-    /// Keep track of peptide IDs and use them for matching?
-    bool use_IDs_;
-
-    /// Whether current cluster is valid
-    bool valid_;
-
+    typedef NeighborMap ClusterElementsMap;
     /** 
-     * @brief Whether initial collection of all neighbors is needed
-     *
-     * This variable stores whether we need to collect all annotations first
-     * before we can decide upon the best set of cluster points. This is
-     * usually only necessary if the center point does not have an annotation
-     * but we want to use ids.
-     *
-    */
-    bool collect_annotations_;
-
-    /// Whether current cluster is accepting new elements or not (if true, no more new elements allowed)
-    bool finalized_;
-
-    /// x coordinate in the grid cell
-    Int x_coord_;
-
-    /// y coordinate in the grid cell
-    Int y_coord_;
-
-    /**
-     * @brief Set of annotations of the cluster
-     *
-     * The set of peptide sequences that is compatible to the cluster center
-     * and results in the best cluster quality.
-     */
-    std::set<AASequence> annotations_;
-
-    /// Base constructor (not accessible)
-    QTCluster();
-
-    /// Computes the quality of the cluster
-    void computeQuality_();
-
-    /**
-     * @brief Finds the optimal annotation (peptide sequences) for the cluster
-     *
-     * The optimal annotation is the one that results in the best quality. It
-     * is stored in @p annotations_;
-     *
-     * This function is only needed when peptide ids are used and the current
-     * center point does not have any peptide id associated with it. In this
-     * case, it is not clear which peptide id the current cluster should use.
-     * The function thus iterates through all possible peptide ids and selects
-     * the one producing the best cluster.
-     *
-     * This function needs access to all possible neighbors for this cluster
-     * and thus can only be run when tmp_neighbors_ is filled (which is during
-     * the filling of a cluster). The function thus cannot be called after
-     * finalizing the cluster.
-     *
-     * @returns The total distance between cluster elements and the center.
+     * @brief Class to store the bulk internal data (neihgbors, annotations, etc.)
+     * 
+     * Has no functionality without a QTCluster (-head) pointing to it.
+     * Create object of this class before calling constructor of QTCluster
      */
-    double optimizeAnnotations_();
-
-public:
+    class OPENMS_DLLAPI Data_
```

Comment:
The best way here would be to make `Data_` private and a struct. 

But we had to make it public, because we wanted to preallocate the `Data_`-Objects inside a vector before we called the constructor of the actual `QTCluster`. We tried to optimize cache behaviour by this while implementing the hot-cold-splitting. 

Since the heap brought such a massive speed-up and made all previous optimizations irrelevant, we could now try to implement the better style version. I don't believe this will affect the runtime and will test it. This would definitely be an improvement of code quality.

Reply:
Made it a struct. Name is `BulkData` now.